helix-core/src/main/java/org/apache/helix/tools/ClusterVerifiers/ZkHelixClusterVerifier.java

```diff
@@ -104,15 +112,49 @@ public ZkHelixClusterVerifier(String zkAddr, String clusterName) {
     if (zkAddr == null || clusterName == null) {
       throw new IllegalArgumentException("requires zkAddr|clusterName");
     }
-    _zkClient = DedicatedZkClientFactory.getInstance()
-        .buildZkClient(new HelixZkClient.ZkConnectionConfig(zkAddr));
+
+    RealmAwareZkClient zkClient;
+    try {
+      // First, try to create a RealmAwareZkClient that's a DedicatedZkClient
+      RealmAwareZkClient.RealmAwareZkConnectionConfig.Builder connectionConfigBuilder =
+          new RealmAwareZkClient.RealmAwareZkConnectionConfig.Builder();
+      connectionConfigBuilder.setZkRealmShardingKey("/" + clusterName);
+      connectionConfigBuilder.setRealmMode(RealmAwareZkClient.RealmMode.SINGLE_REALM);
+      zkClient = DedicatedZkClientFactory.getInstance()
+          .buildZkClient(connectionConfigBuilder.build(),
+              new RealmAwareZkClient.RealmAwareZkClientConfig());
+    } catch (IOException | InvalidRoutingDataException | IllegalStateException e) {
+      // Fallback option is to create a HelixZkClient that's a DedicatedZkClient
```

Comment:
Question, In which case we need to handle this fallback logic? It is expected in some condition or just for safe?

Reply:
Cases where there is no proper MSDS or routing data available. Then we fall back to the previously-existing behavior.